lib/alertStateMachine.js

```diff
@@ -0,0 +1,55 @@
+const ALERT_STATE = require('./alertStateEnum.js')
+
+class AlertStateMachine {
+    // asksIncidentDetails (Boolean)
+    //    true if the user should be asked about the incident details.
+    // getReturnMessage (Function)
+    //    Takes two parameters: the current AlertState and the next AlertState
+    //    Returns the message to send to the user for that state transition
+    constructor (asksIncidentDetails, getReturnMessage) {
+        this.asksIncidentDetails = asksIncidentDetails
+        this.getReturnMessage = getReturnMessage
+    }
+
+    processStateTransitionWithMessage(currentAlertState, messageText, validIncidentCategories) {
+        let nextAlertState
+        let incidentCategory
+        let details
+
+        switch (currentAlertState) {
+            case ALERT_STATE.STARTED:
+            case ALERT_STATE.WAITING_FOR_REPLY:
+                nextAlertState = ALERT_STATE.WAITING_FOR_CATEGORY
+                break
+
+            case ALERT_STATE.WAITING_FOR_CATEGORY:
+                if (messageText.trim() in validIncidentCategories) {
+                    incidentCategory = messageText.trim()
+                    nextAlertState = this.asksIncidentDetails ? ALERT_STATE.WAITING_FOR_DETAILS : nextAlertState = ALERT_STATE.COMPLETED
+                } else {
+                    nextAlertState = currentAlertState
```

Comment:
if the user entered an invalid incident category - how are we informing them by text message that they need to enter a valid one?

Reply:
I was imagining it so that the client (ODetect or Buttons)'s implementation of `getReturnMessage` would also handle the case where the `fromAlertState` and the `toAlertState` are both `WAITING_FOR_CATEGORY`.

Here's where I'm doing it for ODetect: https://github.com/bravetechnologycoop/ODetect-Backend-Local/blob/td-use-alert-lib/index.js#L142